src/core/Akka/IO/TcpListener.cs

```diff
@@ -119,27 +113,74 @@ private Receive Unbinding(IActorRef requester) => message =>
-                
+
                 case Status.Failure fail:
                     _log.Error(fail.Cause, "Failed to unbind TCP listener for address [{0}]", _bind.LocalAddress);
                     Context.Stop(Self);
                     return true;
-                
+
                 default:
                     return false;
             }
         };
-        
-        private IEnumerable<SocketAsyncEventArgs> Accept(int limit)
+
+        private void StartAccept(SocketAsyncEventArgs saea)
         {
-            for(var i = 0; i < limit; i++)
+            var pending = _socket.AcceptAsync(saea);
+            if (!pending)
+                Self.Tell(new AcceptCompleted(saea), Self); // synchronous completion ➔ mailbox
+        }
+
+        private static void OnIoCompleted(object? sender, SocketAsyncEventArgs saea)
+        {
+            // Marshall back into the actor context – keeps user code off the IOCP thread.
+            var actorArgs = (SocketAsyncActorEventArgs)saea;
+
+            var actor = actorArgs.NotifyMe;
+            if (actorArgs.LastOperation == SocketAsyncOperation.Accept)
             {
-                var self = Self;
-                var saea = new SocketAsyncEventArgs();
-                saea.Completed += (_, e) => self.Tell(new SocketEvent(e));
-                if (!_socket.AcceptAsync(saea))
-                    Self.Tell(new SocketEvent(saea));
-                yield return saea;
+                actor.Tell(new AcceptCompleted(saea), actor);
+            }
+            else // should never happen
+            {
+                // This should never happen, but just in case.
+                var ioe = new InvalidOperationException(
+                    $"SocketAsyncEventArgs last operation is not Accept: {actorArgs.LastOperation}");
+                actor.Tell(new Status.Failure(ioe), actor);
+
+                // retry the operation
+                actorArgs.AcceptSocket = null;
+                actor.Tell(new RetryAccept(actorArgs));
+            }
+        }
+
+        private void HandleAccept(SocketAsyncEventArgs saea)
+        {
+            switch (saea.SocketError)
+            {
+                case SocketError.Success:
+                    var accepted = saea.AcceptSocket!;
+                    saea.AcceptSocket = null; // ready for re‑use
+                    Context.ActorOf(Props
+                        .Create<TcpIncomingConnection>(_tcp, accepted, _bind.Handler, _bind.Options, _bind.PullMode)
+                        .WithDeploy(Deploy.Local));
+                    StartAccept(saea); // keep the pool full
+                    break;
+
+                case SocketError.ConnectionReset:
+                case SocketError.NoBufferSpaceAvailable:
+                case SocketError.TryAgain:
+                case SocketError.TimedOut:
+                case SocketError.WouldBlock:
+                    // transient – short back‑off then retry
+                    saea.AcceptSocket = null;
```

Comment:
All of these error codes a retriable - schedule a short back-off and go again.

Important note: can't use `IWithTimers` for this because you'd need a unique timer key each time - hence why I'm using the `IScheduler`. That's something we may want to improve or add to the `IWithTimers` interface.